### DISCUSSION TITLE
How to set bulk parameters via MQTT

### DISCUSSION THREAD

**Discussion by jtronicus** [zwave-js/zwave-js-ui#2851] [Q&A] [answered]
I have a device that I am trying to set 7 sequential parameters all at once using setBulk over MQTT, but so far I have been unsuccessful.

This is what I am trying:

MQTT Topic:
`zwavejs/_CLIENTS/ZWAVE_GATEWAY-zwavejs2mqtt/api/sendCommand/set`

MQTT Payload:
```
{ 
    "args": [
        {
            "nodeId": 69,
            "commandClass": 112,
            "endpoint": 0        },
        "setBulk",
        [
            {"parameter": 60, "value": 54331904},
            {"parameter": 61, "value": 54331932},
            {"parameter": 62, "value": 54331948},
            {"parameter": 63, "value": 54331994},
            {"parameter": 64, "value": 54332031},
            {"parameter": 65, "value": 54332079},
            {"parameter": 66, "value": 54332121}
        ]
  ]
}
```

And this is the error I receive:
```
{
  "success": false,
  "message": "values has the wrong type (ZW0322)",
  "args": [
    {
      "nodeId": 69,
      "commandClass": 112,
      "endpoint": 0
    },
    "setBulk",
    [
      {
        "parameter": 60,
        "value": 54331904
      },
      {
        "parameter": 61,
        "value": 54331932
      },
      {
        "parameter": 62,
        "value": 54331948
      },
      {
        "parameter": 63,
        "value": 54331994
      },
      {
        "parameter": 64,
        "value": 54332031
      },
      {
        "parameter": 65,
        "value": 54332079
      },
      {
        "parameter": 66,
        "value": 54332121
      }
    ]
  ],
  "origin": {
    "args": [
      {
        "nodeId": 69,
        "commandClass": 112,
        "endpoint": 0
      },
      "setBulk",
      [
        {
          "parameter": 60,
          "value": 54331904
        },
        {
          "parameter": 61,
          "value": 54331932
        },
        {
          "parameter": 62,
          "value": 54331948
        },
        {
          "parameter": 63,
          "value": 54331994
        },
        {
          "parameter": 64,
          "value": 54332031
        },
        {
          "parameter": 65,
          "value": 54332079
        },
        {
          "parameter": 66,
          "value": 54332121
        }
      ]
    ]
  }
}
```

Based on the error message I assume values array is not correct, but I have not been able to find any documentation that shows me how the payload should be formatted.  Does anyone know what I am doing wrong?

**Comment by kpine** [comment]
What is the device?

https://devices.zwave-js.io

**Reply by jtronicus** [reply]
The device doesnt exist on https://devices.zwave-js.io/ (its a brand new device that isnt being sold yet).

It does support configuration v4, and parameters 60-66 are all 4-byte unsigned if that helps.

However, I get a similar error when attempting to bulk set parameters on an Inovelli LZW45 LED Strip (its the only other device I own that supports configuration v4)

Topic:
`zwavejs/_CLIENTS/ZWAVE_GATEWAY-zwavejs2mqtt/api/sendCommand/set`

Payload:
```
{ 
    "args": [
        {
            "nodeId": 48,
            "commandClass": 112,
            "endpoint": 0        },
        "setBulk",
        [
            {"parameter": 4, "value": 1},
            {"parameter": 5, "value": 99}
        ]
  ]
}
```

Error:
```
{
  "success": false,
  "message": "values has the wrong type (ZW0322)",
  "args": [
    {
      "nodeId": 48,
      "commandClass": 112,
      "endpoint": 0
    },
    "setBulk",
    [
      {
        "parameter": 4,
        "value": 1
      },
      {
        "parameter": 5,
        "value": 99
      }
    ]
  ],
  "origin": {
    "args": [
      {
        "nodeId": 48,
        "commandClass": 112,
        "endpoint": 0
      },
      "setBulk",
      [
        {
          "parameter": 4,
          "value": 1
        },
        {
          "parameter": 5,
          "value": 99
        }
      ]
    ]
  }
}
```

**Reply by kpine** [reply]
> The device doesnt exist on [devices.zwave-js.io](https://devices.zwave-js.io/) (its a brand new device that isnt being sold yet).

That makes sense then. You need to set the `valueSize` for unknown parameters.

> However, I get a similar error when attempting to bulk set parameters on an Inovelli LZW45 LED Strip (its the only other device I own that supports configuration v4)

Hmm, that part doesn't make sense though, the parameters are defined and the values are in range.

**Comment by kpine** [comment, marked as answer]
`setBulk` takes an array itself, so you need another `[]`:

```json
{ 
    "args": [
        {
            "nodeId": 48,
            "commandClass": 112,
            "endpoint": 0        },
        "setBulk",
        [
          [
            {"parameter": 4, "value": 1},
            {"parameter": 5, "value": 99}
          ]
        ]
  ]
}
```

MQTT is expecting an array of function arguments. In this case, there's one argument which is also an array. If you were using a different API with multiple arguments (e.g. [MultilevelSwitch set](https://zwave-js.github.io/node-zwave-js/#/api/CCs/MultilevelSwitch?id=set)), it would work with just the outer `[]`.

**Reply by jtronicus** [reply]
That solved the problem.  Thanks!

**Comment by robertsLando** [comment]
Thanks @kpine
